### PR TITLE
[RFR] Convert DataProvider to Object

### PR DESCRIPTION
**Pull request by djhi**
Related to #3686

- [x] Add TypeScript type for DataProvider methods
- [x] Make `useDataProvider` return a proxied `dataProvider`
- [x] Remove handling of event in `useMutation` callback
- [x] Add converter for legacy (function) style dataProviders to make the change BC
- [x] Update examples
- [x] Update documentation
  - [x] DataProvider
  - [x] Querying the API
  - [x] Other chapters
- [x] Update core data providers
  - [x] `ra-data-fakerest`
  - [x] `ra-data-simple-rest`
  - [x] `ra-data-json-server`
  - [ ] ~~`ra-data-graphql-simple`~~ (too much work, the object/function wrapper is enough)
  - [ ] ~~`ra-data-graphcool`~~ (too much work, the object/function wrapper is enough)

### CODE REVIEW

File: docs/CreateEdit.md [hunk 1]
```diff
@@ -945,3 +945,106 @@ export const UserEdit = ({ permissions, ...props }) =>
...
+        form.change('average_note', 10);
+
+        handleSubmitWithRedirect('edit');
+    }, [form]);
+
+    return <SaveButton {...props} handleSubmitWithRedirect={handleClick} />;
+};
+```
+
+2. Using react-admin hooks to run custom mutations
```
Comment (djhi) [inline docs/CreateEdit.md]: Note that this will not **trigger** a form validation pass

File: docs/DataProviders.md [hunk 2]
```diff
@@ -5,31 +5,54 @@ title: "Data Providers"
...
+dataProvider
+    .getOne('posts', { id: 123 })
+    .then(response => {
+        console.log(response.data); // { id: 123, title: "hello, world" }
+    })
+```
+
+It's the Data Provider's job to turns these method calls into HTTP requests, and transform the HTTP responses to the data format expected by react-admin. In technical terms, a Data Provider is an *adapter* for an API. 
+
+And to inject a Data Provider in a react-admin application, just pass it as the `dataProvider` prop of the `<Admin>` component, as follows:
```
Comment (djhi) [inline docs/DataProviders.md]:
```suggestion
And to inject a Data Provider in a react-admin application, pass it as the `dataProvider` prop of the `<Admin>` component, as follows:
```

File: docs/DataProviders.md [hunk 3]
```diff
@@ -167,320 +183,170 @@ Now all the requests to the REST API will contain the `X-Custom-Header: foobar`
...
- *  => Promise.resolve({ data: { id: 123, title: "hello, world" } })
- *
- * @param {string} type Request type, e.g GET_LIST
- * @param {string} resource Resource name, e.g. "posts"
- * @param {Object} payload Request parameters. Depends on the action type
- * @returns {Promise} the Promise for a response
- */
-const dataProvider = (type, resource, params) => new Promise();
-```
+APIs are so diverse that quite often, none of the available Data Providers suit you API. In such cases, you'll have to write your own Data Provider. Don't worry, it usually takes a couple hours and presents no difficulty.
```
Comment (djhi) [inline docs/DataProviders.md]:
```suggestion
APIs are so diverse that quite often, none of the available Data Providers suit you API. In such cases, you'll have to write your own Data Provider. Don't worry, it usually takes only a couple hours. 
```

**Comment by fzaninotto** [comment]
Switching to RFR

**Comment by djhi** [comment]
Sorry, need a rebase

File: packages/ra-core/src/dataFetchActions.ts [hunk 4]
```diff
@@ -20,3 +20,28 @@ export const fetchActionsWithArrayOfRecordsResponse = [
     DELETE_MANY,
 ];
 export const fetchActionsWithTotalResponse = [GET_LIST, GET_MANY_REFERENCE];
+
+export const sanitizeFetchType = (fetchType: string) => {
```
Comment (pemoreau) [inline packages/ra-core/src/dataFetchActions.ts]: can't we use types for `GET_LIST` and `getList`, ... this would make switch/case/ safer.
I could try to do it in another PR

File: packages/ra-core/src/dataProvider/useMutation.ts [hunk 5]
```diff
@@ -129,11 +178,131 @@ const useMutation = (
...
+    if (callTimeQuery instanceof Event)
+        return {
+            type: query.type,
+            resource: query.resource,
+            payload: query.payload,
+            options: sanitizeOptions(options),
+        };
+    if (query)
+        return {
+            type: query.type || callTimeQuery.type,
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/dataProvider/useMutation.ts]: No, I don't think so: If both `query` and `callTimeQuery` are falsy, the test line 258 will throw an error. So in this line, at least one of the two is of type `Mutation`, which has a required `type`.

File: packages/ra-core/src/dataProvider/useMutation.ts [hunk 6]
```diff
@@ -129,11 +178,131 @@ const useMutation = (
...
+        payload: callTimeQuery.payload,
+        options: sanitizeOptions(callTimeOptions),
+    };
+};
+
+const hasDeclarativeSideEffectsSupport = (
+    options?: MutationOptions,
+    callTimeOptions?: MutationOptions
+) => {
+    if (!options && !callTimeOptions) return false;
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/dataProvider/useMutation.ts]: I don't think so. Can you elaborate?

File: packages/ra-core/src/dataFetchActions.ts (hunk 4 above)
Comment (fzaninotto) [reply packages/ra-core/src/dataFetchActions.ts]: Well, the types should ba 'open' in that case, because users may pass a custom fetch action (like `placeOrder` or `login`) and we don't want to fail at compile time in this case. 

Can be tested on another PR.

File: packages/ra-core/src/dataProvider/useMutation.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (pemoreau) [reply packages/ra-core/src/dataProvider/useMutation.ts]: May be I do not understand something, but we can have `query !== null` and `query.type` falsy. And in this case, since ` callTimeQuery` may be null, a null exception could occur.

File: packages/ra-core/src/dataProvider/useMutation.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pemoreau) [reply packages/ra-core/src/dataProvider/useMutation.ts]: when `(!options && !callTimeOptions)` is true, none of the two following tests can be true (since the need at least `callTimeOptions`  or `options` to exist)

Therefore, the last `return false` is sufficient. 
In other words : by default `false` is returned, except when `(callTimeOptions && callTimeOptions.withDeclarativeSideEffectsSupport)` or `(options && options.withDeclarativeSideEffectsSupport)`

File: packages/ra-core/src/dataFetchActions.ts (hunk 4 above)
Comment (pemoreau) [reply packages/ra-core/src/dataFetchActions.ts]: OK I understand.
